README.md
# Shepard5 NFL Model
The goal of this repo is to identify favorable week to week nfl spreads, win/loss, over/under lines such that in the long term a sportsbettor will return a positive rate of return.
The repo consists primarily of data collection modules along with R scripts that tinker data extracted from the python functions. 
I suggest you refer to the TeamID excel sheet to link NFL teams with their corresponding indices in the automatically generated excel sheets.


#
Developing models for individual teams rather than using all available data to summarize into one singular model will improve predictability game outcomes. For example, QB-centric teams (KC, BAL, Bills, Rams) will find that QBR is a better predictor variable than teams with a more homogeneous offense (49ers, Patriots, Steelers, Titans). 

I ran a quick linear regression in R for QB-centric (QB heavy) (BUF,NYJ,CIN,BAL,JAX,KC,DEN,PHI,CHI,ARI) and QB-non-centric (QB light) (NE,CLE,PIT,HOU,TEN,LV,WSH,DET,SEA,49ers) teams. I also balanced it with sacks to account for variations on defensive performances. While it is a small sample size, the results support our intuition. 


![Screen Shot 2023-08-24 at 9 40 55 PM](https://github.com/shepard5/NFL/assets/108085853/884cc169-580f-4c4e-8da4-95a0c6e5c491)




#
Introducing logistic regression model to predict over/under and moneyline outcomes. Teams will be selected based on predictability - If a team's roster has changed heavily since last season, will have to wait until data comes out from the first few weeks to make decisions. Need to make a detailed list of teams whose rosters are highly similar or differnet compared to the 2022 season; factors to consider - young players improving, older players depreciating, coaching changes for struggling teams, qb changes, maybe consider changes that have been overcompensated (Falcons - Bijon Robinson, Desmond Ridder ?). Teams with high predictability will be selected for specialized model development. Will then look for identical matchups (or close to) present in both 2022 and 2023 between teams with high (and possibly low predictability) and run regression using 2022 season data. 


Changelog
8/15
Found data pipeline (ESPN hidden API) provides ats information in real time as well as matchup information and scores. Need to combine all functions into a single functions.py file for ease of use. Need to incorporate other predictor variables into the model and perform some kind of regression analysis to review performance retrofitting historical data and verifying predictability.

8/16
Produced a script that generates QBR for individual teams - need to be able to run for all teams at once and output to excel sheet. Later today will do an assessment of which tasks need to be prioritized. 

8/17
Trying to figure out how to get the dataset formatted well. Filling zeros for teams with byes and NFC AFC. Removing repeated values if 2 qbs play for one week. Combined functions into one py script to export all data to an excel sheet. (Changes/notes that were put on readme file 8/17) 'As of 8/17 - Focusing on developing scripts that can seamlessly transmit predicting variables to an excel file to be imported to R Studio. Seamlessly enough to circumstantially produce datasheets that will be interpreted by R Studio week to week without interruption - fluctuating bye-weeks, AFC NFC mismatches, multiple inputs of a statistic when only one is expected, and all other unknown unexpected results - should be accounted for and blended into each pipeline.'

8/18
Solved bye week and AFC NFC issue with QBR data. Consolidated functions into one. Simplified excel export, but may want to explore how openpyxl interprets different objects - it has a nonintuitive way of delivering keys and tuple values to a dataframe separately. Using dictionaries homogeneously throughout future data retrieval functions will make formatting and exporting 10x easier. Also need to look deeper into Kirk Cousins / Mullen QBR selection and output. 

If this changelog stops being a changelog it's because I figured out how to use GitHub



All_Functions.py
import requests
import pandas as pd
import openpyxl

def NFL_ATS_to_Excel():

    streak_data = []
    
    for x in range(1,35):
        if x in [31,32]:
            continue
        
        ATS_streak = NFL_ATS(str(x))
        streak_data.append([x, ATS_streak])
            
    print(streak_data)
    
    df = pd.DataFrame(streak_data, columns=['TeamID', 'Streak (2 or more ATS)'])

    with pd.ExcelWriter('ATS_Streak_output.xlsx', engine='openpyxl') as writer:
        df.to_excel(writer, index = False) 
    








def NFL_ATS():

    #See ESPN TeamID file for reference
    
    #response = requests.get('https://sports.core.api.espn.com/v2/sports/football/leagues/nfl/seasons/' + year + '/types/2/teams/' + team_id + '/ats')
    #ATS_totals = response.json()

    ats_data = [] 
    for x in range(1, 35):
        if x == 31 or x == 32:
            prospect = ""
            ats_data.append([x,prospect])
            continue
        team_url = f'https://sports.core.api.espn.com/v2/sports/football/leagues/nfl/teams/{x}/odds/1002/past-performances?limit=300'
        response = requests.get(team_url)
        ATS_historical = response.json()

        week_recent = ATS_historical["items"][len(ATS_historical['items'])-1]
        week_2nd_recent = ATS_historical["items"][len(ATS_historical['items'])-2]

        #print(week_recent)
        #print(week_2nd_recent)
        print(team_url)
        print(week_recent["spreadWinner"])
        if week_recent["spreadWinner"] == week_2nd_recent["spreadWinner"]: #If a team is on an ATS win or lose streak prospect will return as hot or cold respectively.
            if week_recent["spreadWinner"] == True:
                prospect = 'hot'
            else:
                prospect = 'cold'
        else:
            prospect = ""
        
        ats_data.append([x,prospect])
    return ats_data

    







def QBR():


#Find most recently played week for most recent QBR data
    
    for x in range(23,21,-1):
        for y in range(18,0,-1):
            url = f"https://sports.core.api.espn.com/v2/sports/football/leagues/nfl/seasons/20{x}/types/2/weeks/{y}/qbr/10000?limit=300"       
            headers = {
    
            }

            response = requests.get(url, headers=headers)
            QBR_data = response.json()
            
            Team_QBR = 0
        
            if len(QBR_data["items"])>1:

                url = f"https://sports.core.api.espn.com/v2/sports/football/leagues/nfl/seasons/20{x}/types/2/weeks/{y}/qbr/10000?limit=300"
                headers = {
    
                }

                response = requests.get(url, headers=headers)

                QBR_data = response.json()
                Team_QBR = 1

                year = f"20{x}"
                week = y
                
                break       

        if Team_QBR == 1:    
            break

 #  excel_file_path = "C:\Users\Sam\Desktop\NFL-main\Script Outputs ETC\ATS_Streak_output.xlsx"
    sheet_name = "Sheet1"

    data = [] 
    #for x in range(1,35):
    #    data.append(x) 
    print(data)
    seen = set()
    
    for x in range(0,len(QBR_data["items"])-1): 

        Team_QBR = QBR_data["items"][x]["splits"]["categories"][0]["stats"][17]["value"]
        Team_ID_url = QBR_data["items"][x]["team"]["$ref"]
        Team_ID = ""

        if Team_ID_url[83] != "?":
            Team_ID = int(Team_ID_url[82] + Team_ID_url[83])

        else:
            Team_ID = int(Team_ID_url[82])

        if Team_ID not in seen and x<35 :
            seen.add(Team_ID)
            data.append([Team_ID, Team_QBR])
    print(data)

    
    sorted_data = sorted(data, key=lambda x: x[0])
    #print(sorted_data)

    TeamID_in_data = [sublist[0] for sublist in data]

    for x in range (1,35):
        if x not in TeamID_in_data:
            data.append([x,0])

    sorted_data = sorted(data, key=lambda x: x[0])
    print(sorted_data)


    file_path = r"C:\Users\Sam\Desktop\NFL-main\Team_Data.xlsx"
    sheet_name = 'Sheet1'  # Adjust this to your sheet's name if different

    df = pd.DataFrame(sorted_data)
    df.columns = ["Team ID","QBR"]

    
    with pd.ExcelWriter(file_path, engine='openpyxl') as writer:
        df.to_excel(writer, sheet_name='Sheet1', index=False)






##8/17 Notes - Trying to figure out how to get the data set formatted well. Filling in zeros for teams with byes and NFC AFC. Removing repeated players if a qb gets subbed in for the week. 



#Pull QBR data for desired team from Team_ID and season / week we just gathered
        
#    for x in range(0,len(QBR_data["items"])-1): 
#        correct_team = f"http://sports.core.api.espn.com/v2/sports/football/leagues/nfl/seasons/{year}/teams/{Team_ID}?lang=en&region=us"
# 
#        if correct_team == str(QBR_data["items"][x]["team"]["$ref"]):
#            Team_QBR = QBR_data["items"][x]["splits"]["categories"][0]["stats"][17]["value"]
#            return(Team_QBR)


## For future use
#   return year
#   return week
